core/src/main/java/com/linecorp/armeria/common/HttpRequestBuilder.java

```diff
@@ -0,0 +1,397 @@
+     */
+    public HttpRequestBuilder cookie(Cookie cookie) {
+        cookies.add(requireNonNull(cookie, "cookie"));
+        return this;
+    }
+
+    /**
+     * Sets multiple cookies for this request. For example:
+     * <pre>{@code
+     * HttpRequest.builder()
+     *            .get("/")
+     *            .cookies(Cookies.of(Cookie.of("cookie1", "foo"),
+     *                                Cookie.of("cookie2", "bar")))
+     *            .build();
+     * }</pre>
+     * @see Cookies
+     */
+    public HttpRequestBuilder cookies(Cookies cookies) {
+        this.cookies.addAll(requireNonNull(cookies, "cookies"));
+        return this;
+    }
+
+    /**
+     * Creates a new {@link HttpRequest}.
+     */
+    public HttpRequest build() {
+        final RequestHeaders requestHeaders = requestHeaders();
+        if (content == null || content.isEmpty()) {
+            if (content != null) {
+                content.close();
+            }
+            if (httpTrailersBuilder.isEmpty()) {
+                return new EmptyFixedHttpRequest(requestHeaders);
+            }
+            return new OneElementFixedHttpRequest(requestHeaders, httpTrailersBuilder.build());
+        }
+        if (httpTrailersBuilder.isEmpty()) {
+            return new OneElementFixedHttpRequest(requestHeaders, content);
+        }
+        return new TwoElementFixedHttpRequest(requestHeaders, content, httpTrailersBuilder.build());
+    }
+
+    private RequestHeaders requestHeaders() {
+        requestHeadersBuilder.path(buildPath());
+        if (!cookies.isEmpty()) {
+            requestHeadersBuilder.set(COOKIE, Cookie.toCookieHeader(cookies));
+        }
+        if (content == null || content.isEmpty()) {
+            requestHeadersBuilder.remove(CONTENT_LENGTH);
+        } else {
+            requestHeadersBuilder.setInt(CONTENT_LENGTH, content.length());
+        }
+        return requestHeadersBuilder.build();
+    }
+
+    private String buildPath() {
+        checkState(path != null, "path must be set.");
+        final StringBuilder pathBuilder = new StringBuilder(path);
+        if (!pathParams.isEmpty()) {
+            int i = 0;
+            while (i < pathBuilder.length()) {
+                if (pathBuilder.charAt(i) == '{') {
+                    int j = i;
+                    while (j < pathBuilder.length() && pathBuilder.charAt(j) != '}') {
+                        j++;
+                    }
+                    if (j == pathBuilder.length()) {
+                        break;
+                    }
+                    final String name = pathBuilder.substring(i + 1, j);
+                    if (pathParams.containsKey(name)) {
+                        final String value = pathParams.get(name).toString();
+                        pathBuilder.replace(i, j + 1, value);
+                        i += value.length() - 1;
+                    }
+                } else if (pathBuilder.charAt(i) == ':') {
+                    int j = i;
```

Comment:
Ditto
```suggestion
                    int j = i + 1;
```
And check the end condition.